### PR TITLE
add ability for application webhook

### PR DESCRIPTION
**Pull request by nitro-neal**
# Overview
Add the ability to fire a webhook when someone submits an application.

# Description
To listen to new events when an application is submitted:

```
POST:  - http://localhost:8080/v1/webhooks
{
    "noun": "Application",
    "verb": "Create",
    "url": "http://host.docker.internal:8081/webhook"
}
```

When an application is submitted against these will fire from the webhook:

```
{
  "noun": "Application",
  "verb": "Create",
  "url": "http://host.docker.internal:8081/webhook",
  "data": {
    "id": "credentials/responses/id123",
    "done": false,
    "result": {}
  }
} 
```

```
 {
  "noun": "Application",
  "verb": "Create",
  "url": "http://host.docker.internal:8081/webhook",
  "data": {
    "id": "credentials/responses/id123",
    "done": true,
    "result": {
      "response": {
        "credential_response": {
          "id": "c31ee2b3-1825-41b5-b882-62700f850c85",
          "spec_version": "https://identity.foundation/credential-manifest/spec/v1.0.0/",
          "manifest_id": "e46c457f-3fc5-4500-89d5-b278799af475",
          "application_id": "id123",
          "fulfillment": {
            "descriptor_map": [
              {
                "id": "c9f506e3-8752-4139-a862-3b831c517cf8",
                "format": "jwt_vc",
                "path": "$.verifiableCredentials[0]"
              }
            ]
          }
        },
        "verifiableCredentials": [
          "eyJhbGciOiJFZERTQSIsImtpZCI6IiN6Nk1raTZxdGEzSmhxM3JGVXNDNHF5WDNad05tU3hTc0xQSDY1YmlmN2tCeDJrVWUiLCJ0eXAiOiJKV1QifQ.eyJleHAiOjE2NjcxNzQ0MDAsImlhdCI6MTY4MzY1MzA4MiwiaXNzIjoiZGlkOmtleTp6Nk1raTZxdGEzSmhxM3JGVXNDNHF5WDNad05tU3hTc0xQSDY1YmlmN2tCeDJrVWUiLCJqdGkiOiJjOWY1MDZlMy04NzUyLTQxMzktYTg2Mi0zYjgzMWM1MTdjZjgiLCJuYmYiOjE2ODM2NTMwODIsIm5vbmNlIjoiNDIyZGYxMTktNTlhNC00OGEyLWI3ZjItYTQ4MTVhMjE3YTgwIiwic3ViIjoiZGlkOmtleTp6Nk1rcEd4NVJiNlpwR0RQaGR1OHlGb3cyR3prd1g0VHZqSG56a3BGd2ZjWThzRG0iLCJ2YyI6eyJAY29udGV4dCI6WyJodHRwczovL3d3dy53My5vcmcvMjAxOC9jcmVkZW50aWFscy92MSJdLCJ0eXBlIjpbIlZlcmlmaWFibGVDcmVkZW50aWFsIl0sImlzc3VlciI6IiIsImlzc3VhbmNlRGF0ZSI6IiIsImNyZWRlbnRpYWxTdGF0dXMiOnsiaWQiOiJodHRwOi8vbG9jYWxob3N0OjgwODAvdjEvY3JlZGVudGlhbHMvYzlmNTA2ZTMtODc1Mi00MTM5LWE4NjItM2I4MzFjNTE3Y2Y4L3N0YXR1cyIsInN0YXR1c0xpc3RDcmVkZW50aWFsIjoiaHR0cDovL2xvY2FsaG9zdDo4MDgwL3YxL2NyZWRlbnRpYWxzL3N0YXR1cy8xM2IyMTJmYi1mODk3LTRlY2QtYTdkYS04YWUwOWIzYTAyNjkiLCJzdGF0dXNMaXN0SW5kZXgiOiIzNTgxMCIsInN0YXR1c1B1cnBvc2UiOiJyZXZvY2F0aW9uIiwidHlwZSI6IlN0YXR1c0xpc3QyMDIxRW50cnkifSwiY3JlZGVudGlhbFN1YmplY3QiOnsiZmFtaWx5TmFtZSI6InNpbXBzb24iLCJnaXZlbk5hbWUiOiJyaWNreSBib2JieSIsInBvc3RhbEFkZHJlc3MiOnsiYWRkcmVzc1JlZ2lvbiI6IkNBIn19LCJjcmVkZW50aWFsU2NoZW1hIjp7ImlkIjoiMmUxYjU1YjQtNTJiMy00ZjRlLTkzY2UtZGRhZTVkZjAyY2E1IiwidHlwZSI6Ikpzb25TY2hlbWFWYWxpZGF0b3IyMDE4In19fQ.RJ-uy8-gZBma0Yu8CZYsLByxP8z0eRuqaN50AcG9KkpIm2z8heCE_FSjVu1LNPhrPOYfLadOUaCPa1QBGAmnDA"
        ],
        "responseJwt": "eyJhbGciOiJFZERTQSIsImtpZCI6IiN6Nk1raTZxdGEzSmhxM3JGVXNDNHF5WDNad05tU3hTc0xQSDY1YmlmN2tCeDJrVWUiLCJ0eXAiOiJKV1QifQ.eyJjcmVkZW50aWFsX3Jlc3BvbnNlIjp7ImFwcGxpY2F0aW9uX2lkIjoiaWQxMjMiLCJmdWxmaWxsbWVudCI6eyJkZXNjcmlwdG9yX21hcCI6W3siZm9ybWF0Ijoiand0X3ZjIiwiaWQiOiJjOWY1MDZlMy04NzUyLTQxMzktYTg2Mi0zYjgzMWM1MTdjZjgiLCJwYXRoIjoiJC52ZXJpZmlhYmxlQ3JlZGVudGlhbHNbMF0ifV19LCJpZCI6ImMzMWVlMmIzLTE4MjUtNDFiNS1iODgyLTYyNzAwZjg1MGM4NSIsIm1hbmlmZXN0X2lkIjoiZTQ2YzQ1N2YtM2ZjNS00NTAwLTg5ZDUtYjI3ODc5OWFmNDc1Iiwic3BlY192ZXJzaW9uIjoiaHR0cHM6Ly9pZGVudGl0eS5mb3VuZGF0aW9uL2NyZWRlbnRpYWwtbWFuaWZlc3Qvc3BlYy92MS4wLjAvIn0sImlhdCI6MTY4MzY1MzA4MiwiaXNzIjoiZGlkOmtleTp6Nk1raTZxdGEzSmhxM3JGVXNDNHF5WDNad05tU3hTc0xQSDY1YmlmN2tCeDJrVWUiLCJ2ZXJpZmlhYmxlQ3JlZGVudGlhbHMiOlsiZXlKaGJHY2lPaUpGWkVSVFFTSXNJbXRwWkNJNklpTjZOazFyYVRaeGRHRXpTbWh4TTNKR1ZYTkROSEY1V0ROYWQwNXRVM2hUYzB4UVNEWTFZbWxtTjJ0Q2VESnJWV1VpTENKMGVYQWlPaUpLVjFRaWZRLmV5SmxlSEFpT2pFMk5qY3hOelEwTURBc0ltbGhkQ0k2TVRZNE16WTFNekE0TWl3aWFYTnpJam9pWkdsa09tdGxlVHA2TmsxcmFUWnhkR0V6U21oeE0zSkdWWE5ETkhGNVdETmFkMDV0VTNoVGMweFFTRFkxWW1sbU4ydENlREpyVldVaUxDSnFkR2tpT2lKak9XWTFNRFpsTXkwNE56VXlMVFF4TXprdFlUZzJNaTB6WWpnek1XTTFNVGRqWmpnaUxDSnVZbVlpT2pFMk9ETTJOVE13T0RJc0ltNXZibU5sSWpvaU5ESXlaR1l4TVRrdE5UbGhOQzAwT0dFeUxXSTNaakl0WVRRNE1UVmhNakUzWVRnd0lpd2ljM1ZpSWpvaVpHbGtPbXRsZVRwNk5rMXJjRWQ0TlZKaU5scHdSMFJRYUdSMU9IbEdiM2N5UjNwcmQxZzBWSFpxU0c1NmEzQkdkMlpqV1RoelJHMGlMQ0oyWXlJNmV5SkFZMjl1ZEdWNGRDSTZXeUpvZEhSd2N6b3ZMM2QzZHk1M015NXZjbWN2TWpBeE9DOWpjbVZrWlc1MGFXRnNjeTkyTVNKZExDSjBlWEJsSWpwYklsWmxjbWxtYVdGaWJHVkRjbVZrWlc1MGFXRnNJbDBzSW1semMzVmxjaUk2SWlJc0ltbHpjM1ZoYm1ObFJHRjBaU0k2SWlJc0ltTnlaV1JsYm5ScFlXeFRkR0YwZFhNaU9uc2lhV1FpT2lKb2RIUndPaTh2Ykc5allXeG9iM04wT2pnd09EQXZkakV2WTNKbFpHVnVkR2xoYkhNdll6bG1OVEEyWlRNdE9EYzFNaTAwTVRNNUxXRTROakl0TTJJNE16RmpOVEUzWTJZNEwzTjBZWFIxY3lJc0luTjBZWFIxYzB4cGMzUkRjbVZrWlc1MGFXRnNJam9pYUhSMGNEb3ZMMnh2WTJGc2FHOXpkRG80TURnd0wzWXhMMk55WldSbGJuUnBZV3h6TDNOMFlYUjFjeTh4TTJJeU1USm1ZaTFtT0RrM0xUUmxZMlF0WVRka1lTMDRZV1V3T1dJellUQXlOamtpTENKemRHRjBkWE5NYVhOMFNXNWtaWGdpT2lJek5UZ3hNQ0lzSW5OMFlYUjFjMUIxY25CdmMyVWlPaUp5WlhadlkyRjBhVzl1SWl3aWRIbHdaU0k2SWxOMFlYUjFjMHhwYzNReU1ESXhSVzUwY25raWZTd2lZM0psWkdWdWRHbGhiRk4xWW1wbFkzUWlPbnNpWm1GdGFXeDVUbUZ0WlNJNkluTnBiWEJ6YjI0aUxDSm5hWFpsYms1aGJXVWlPaUp5YVdOcmVTQmliMkppZVNJc0luQnZjM1JoYkVGa1pISmxjM01pT25zaVlXUmtjbVZ6YzFKbFoybHZiaUk2SWtOQkluMTlMQ0pqY21Wa1pXNTBhV0ZzVTJOb1pXMWhJanA3SW1sa0lqb2lNbVV4WWpVMVlqUXROVEppTXkwMFpqUmxMVGt6WTJVdFpHUmhaVFZrWmpBeVkyRTFJaXdpZEhsd1pTSTZJa3B6YjI1VFkyaGxiV0ZXWVd4cFpHRjBiM0l5TURFNEluMTlmUS5SSi11eTgtZ1pCbWEwWXU4Q1pZc0xCeXhQOHowZVJ1cWFONTBBY0c5S2twSW0yejhoZUNFX0ZTalZ1MUxOUGhyUE9ZZkxhZE9VYUNQYTFRQkdBbW5EQSJdfQ.qOL7U8-1DsrEgXLwPuDZwiHhaTp_Lss3ApCIMFizsg5Rw0VGxEPFAFylGNIb9JPF3zC1I8LCLM_vijBI3neXCg"
      }
    }
  }
} 
```

### CODE REVIEW

File: pkg/server/server.go [hunk 1]
```diff
@@ -246,16 +246,16 @@ func (s *SSIServer) ManifestAPI(service svcframework.Service) (err error) {
...
-	s.Handle(http.MethodPut, manifestHandlerPath, manifestRouter.CreateManifest)
+	s.Handle(http.MethodPut, manifestHandlerPath, manifestRouter.CreateManifest, middleware.Webhook(webhookService, webhook.Manifest, webhook.Create))
 
 	s.Handle(http.MethodGet, manifestHandlerPath, manifestRouter.GetManifests)
 	s.Handle(http.MethodGet, path.Join(manifestHandlerPath, "/:id"), manifestRouter.GetManifest)
-	s.Handle(http.MethodDelete, path.Join(manifestHandlerPath, "/:id"), manifestRouter.DeleteManifest)
+	s.Handle(http.MethodDelete, path.Join(manifestHandlerPath, "/:id"), manifestRouter.DeleteManifest, middleware.Webhook(webhookService, webhook.Manifest, webhook.Delete))
 
-	s.Handle(http.MethodPut, applicationsHandlerPath, manifestRouter.SubmitApplication)
+	s.Handle(http.MethodPut, applicationsHandlerPath, manifestRouter.SubmitApplication, middleware.Webhook(webhookService, webhook.Application, webhook.Create))
```
Comment (andresuribe87) [inline pkg/server/server.go:255]: Should the verb here be submit?

**Review by andresuribe87 — COMMENTED** [review]
Overall this looks good.

Would it make sense to add documentation so that developers know what Noun/Verb tuples are supported for webhooks? I think a good place for this would be the [Create Webhook](https://developer.tbd.website/docs/apis/ssi-service/#tag/WebhookAPI/paths/~1v1~1webhooks/put) description.

Alternatively, each method that supports it could document the Name/Verb for the webhook event.

**Review by decentralgabe — APPROVED** [review]
looks great!
do we have an issue to track webhooks for presentation submissions too?

File: pkg/server/server.go (hunk 1 above)
Comment (nitro-neal) [reply pkg/server/server.go:255]: Yea I was thinking about that, I think its the only thing in the entire system that you 'submit' so submit would only be used for that, and it actually creates a cred app in our system, so thats why I left it as 'create'

**Comment by nitro-neal** [comment]
Added submission:
```
 {
  "noun": "Submission",
  "verb": "Create",
  "url": "http://host.docker.internal:8081/webhook",
  "data": {
    "id": "presentations/submissions/e875b34e-35fd-4ad9-8805-4f16bf98df71",
    "done": false,
    "result": {}
  }
} 
```

File: pkg/server/server.go (hunk 1 above)
Comment (andresuribe87) [reply pkg/server/server.go:255]: My preference is still 'submit', but as long as this is properly documented and explained to external developers then 👌

**Pull request merged**